Implementing configurable aggregation cardinality limit

ThomsonTan opened the pull request:
## Changes

This resumes the PR #3299.

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

ThomsonTan commented on sdk/test/metrics/async_metric_storage_test.cc:
```diff
@@ -25,11 +25,13 @@
 #include "opentelemetry/sdk/metrics/state/metric_collector.h"
-#include "opentelemetry/sdk/metrics/view/attributes_processor.h"
 
 #ifdef ENABLE_METRICS_EXEMPLAR_PREVIEW
+#  include "opentelemetry/sdk/metrics/data/exemplar_data.h"
 #  include "opentelemetry/sdk/metrics/exemplar/filter_type.h"
 #  include "opentelemetry/sdk/metrics/exemplar/reservoir.h"
+#else
+#  include "opentelemetry/sdk/metrics/view/attributes_processor.h"
 #endif
 
```
It was suggested by iwyu. Let's keep it for now.

marcalff reviewed (commented):
Thanks for the follow up on the original PR.

A couple of questions:

1)

About the title:
> Implementing configurable aggregation cardinality limit and collector emitting all views instead of last view

This change seem to be only about "configurable aggregation cardinality limit",
I don't see the relevance with "collector emitting all views instead of last view",
please clarify.

2)

A lot of code checks:

```
aggregation_config
  ? aggregation_config->cardinality_limit_
  : kAggregationCardinalityLimit
```

Would it be simpler to always provide am aggregation config, so that testing for a nullptr is no longer needed, resulting in:

```
aggregation_config->cardinality_limit_
```

dbarker commented on CHANGELOG.md:
```diff
@@ -33,6 +33,10 @@ Increment the:
 * [BUILD] Use -dev versions in main branch
   [#3609](https://github.com/open-telemetry/opentelemetry-cpp/pull/3609)
 
+* [SDK] Implementing configurable aggregation cardinality limit and collector
```
Following up on https://github.com/open-telemetry/opentelemetry-cpp/pull/3624#pullrequestreview-3192770436. The `emitting all views instead of the last view` fix was included in https://github.com/open-telemetry/opentelemetry-cpp/pull/3358 and can be removed from the title and readme in this PR.

dbarker commented on sdk/include/opentelemetry/sdk/metrics/state/sync_metric_storage.h:
```diff
@@ -63,10 +63,12 @@ class SyncMetricStorage : public MetricStorage, public SyncWritableMetricStorage
       : instrument_descriptor_(instrument_descriptor),
-        attributes_hashmap_(new AttributesHashMap(attributes_limit)),
+        aggregation_config_(aggregation_config),
+        attributes_hashmap_(new AttributesHashMap(aggregation_config
```
nit - for creating a unique_ptr we should now use `std::make_unique` instead of new (rationale - https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-make_unique)

ThomsonTan commented:
> Thanks for the follow up on the original PR.
> 
> A couple of questions:
> 
> 
> About the title:
> 
> > Implementing configurable aggregation cardinality limit and collector emitting all views instead of last view
> 
> This change seem to be only about "configurable aggregation cardinality limit", I don't see the relevance with "collector emitting all views instead of last view", please clarify.
> 
> 
> A lot of code checks:
> 
> ```
> aggregation_config
>   ? aggregation_config->cardinality_limit_
>   : kAggregationCardinalityLimit
> ```
> 
> Would it be simpler to always provide am aggregation config, so that testing for a nullptr is no longer needed, resulting in:
> 
> ```
> aggregation_config->cardinality_limit_
> ```

Addressed the feedback:

1. Removed the emitting all views as suggested by @dbarker.
2. Created a default `AggregationConfig`.

lalitb commented on sdk/test/metrics/sum_aggregation_test.cc:
```diff
@@ -142,6 +145,85 @@ TEST(HistogramToSumFilterAttributes, Double)
+        {
+          // Something weird about attributes hashmap. If cardinality is setup to n, it emits n-1
+          // including overflow. Just making the logic generic here to succeed for n or n-1 total
+          // cardinality.
```
This seems to be contradicting the earlier comment 
```
 // Note: When the cardinality limit is set to n, the attributes hashmap emits n-1 distinct


  // attribute sets, plus an overflow bucket for additional attributes. The test logic below is made
  // generic to succeed for either n or n-1 total cardinality. If this behavior is unexpected,
  // please investigate and file an issue.
```

lalitb commented on sdk/test/metrics/sum_aggregation_test.cc:
```diff
@@ -142,6 +145,85 @@ TEST(HistogramToSumFilterAttributes, Double)
+          // including overflow. Just making the logic generic here to succeed for n or n-1 total
+          // cardinality.
+          EXPECT_GE(cardinality_limit, md.point_data_attr_.size());
+          EXPECT_LT(cardinality_limit / 2, md.point_data_attr_.size());
```
Why is this range check done. can't we get exact ?
